JUnit4 test reports contain the wrong test class name

See original conversation here: https://groups.google.com/forum/#!topic/bndtools-users/ebpp8vUiw0A

Test case entries in the JUnit XML report contain a generic test facade class for the test class name such as this entry:

    <testcase name="testAddressCreationWithName" time="0.22" classname="junit.framework.JUnit4TestCaseFacade">
      <error type="java.lang.AssertionError"><![CDATA[
        java.lang.AssertionError: Service is unavailable: xxx.yyy.zzz.SomeService at 
        org.junit.Assert.fail(Assert.java:88) at org.junit.Assert.assertTrue(Assert.java:41) at 
        org.junit.Assert.assertNotNull(Assert.java:621) at
        org.knowhowlab.osgi.testing.assertions.ServiceAssert.assertServiceAvailable(ServiceAssert.java:71) at 
        org.knowhowlab.osgi.testing.assertions.ServiceAssert.assertServiceAvailable(ServiceAssert.java:56) at 
        xxx.yyy.zzz.TestAddressManagerService.setUp(TestAddressManagerService.java:53) at 
        sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at
        sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) at 
        sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at 
        java.lang.reflect.Method.invoke(Method.java:606) at 
        org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47) at 
        org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) at 
        org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44) at 
        org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24) at 
        org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271) at 
        org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70) at 
        org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50) at 
        org.junit.runners.ParentRunner$3.run(ParentRunner.java:238) at 
        org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63) at 
        org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236) at 
        org.junit.runners.ParentRunner.access$000(ParentRunner.java:53) at 
        org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229) at 
        org.junit.runners.ParentRunner.run(ParentRunner.java:309) at 
        junit.framework.JUnit4TestAdapter.run(JUnit4TestAdapter.java:38) at 
        junit.framework.TestSuite.runTest(TestSuite.java:255) at 
        junit.framework.TestSuite.run(TestSuite.java:250) at aQute.junit.Activator.test(Activator.java:302) at 
        aQute.junit.Activator.automatic(Activator.java:185) at aQute.junit.Activator.run(Activator.java:119) at 
        aQute.launcher.Launcher$4.call(Launcher.java:1107) at 
        aQute.launcher.Launcher$4.call(Launcher.java:1105) at 
        aQute.launcher.Launcher.run(Launcher.java:239) at 
        aQute.launcher.Launcher.main(Launcher.java:90)
        ]]>
      </error>
    </testcase>

I would expect classname to be xxx.yyy.zzz.TestAddressManagerService.

To be clear, I would expect the actual Java class containing the test methods to be listed in the classname attribute so it is easier to view test results in a report that contains many different classes. As it is now, it is very difficult and sometimes impossible to figure out which class a test method comes from.

--- a/biz.aQute.junit/src/aQute/junit/JunitXmlReport.java
+++ b/biz.aQute.junit/src/aQute/junit/JunitXmlReport.java
@@ -14,6 +14,8 @@ import java.util.StringTokenizer;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import org.junit.runner.Describable;
+import org.junit.runner.Description;
 import org.osgi.framework.Bundle;
 
 import junit.framework.AssertionFailedError;
@@ -134,10 +136,17 @@ public class JunitXmlReport implements TestReporter {
 		String name = nameAndClass;
 		String clazz = test.getClass().getName();
 
-		Matcher m = NAMEANDCLASS.matcher(nameAndClass);
-		if (m.matches()) {
-			name = m.group(1);
-			clazz = m.group(2);
+		if (test instanceof Describable) {
+			Description description = ((Describable) test).getDescription();
+			clazz = description.getClassName();
+			if (description.getMethodName() != null)
+				name = description.getMethodName();
+		} else {
+			Matcher m = NAMEANDCLASS.matcher(nameAndClass);
+			if (m.matches()) {
+				name = m.group(1);
+				clazz = m.group(2);
+			}
 		}
 
 		testcase = new Tag("testcase");

